pkg/scheduler/objects/queue.go

```diff
@@ -0,0 +1,1009 @@
+	adminACL           security.ACL        // admin ACL
+	submitACL          security.ACL        // submit ACL
+	maxResource        *resources.Resource // When not set, max = nil
+	guaranteedResource *resources.Resource // When not set, Guaranteed == 0
+	allocatedResource  *resources.Resource // set based on allocation
+	isLeaf             bool                // this is a leaf queue or not (i.e. parent)
+	isManaged          bool                // queue is part of the config, not auto created
+	stateMachine       *fsm.FSM            // the state of the queue for scheduling
+	stateTime          time.Time           // last time the state was updated (needed for cleanup)
+
+	sync.RWMutex
+}
+
+func newBlankQueue() *Queue {
+	return &Queue{
+		children:          make(map[string]*Queue),
+		applications:      make(map[string]*Application),
+		reservedApps:      make(map[string]int),
+		properties:        make(map[string]string),
+		stateMachine:      NewObjectState(),
+		allocatedResource: resources.NewResource(),
+		preempting:        resources.NewResource(),
+		pending:           resources.NewResource(),
+	}
+}
+
+// Create a new queue from scratch based on the configuration
+// lock free as it cannot be referenced yet
+func NewConfiguredQueue(conf configs.QueueConfig, parent *Queue) (*Queue, error) {
+	sq := newBlankQueue()
+	sq.Name = strings.ToLower(conf.Name)
+	sq.QueuePath = strings.ToLower(conf.Name)
+	sq.parent = parent
+	sq.isManaged = true
+
+	// update the properties
+	if err := sq.setQueueConfig(conf); err != nil {
+		return nil, fmt.Errorf("configured queue creation failed: %s", err)
+	}
+
+	// add to the parent, we might have an overall lock already
+	// still need to make sure we lock the parent so we do not interfere with scheduling
+	if parent != nil {
+		sq.QueuePath = parent.QueuePath + configs.DOT + sq.Name
+		err := parent.addChildQueue(sq)
+		if err != nil {
+			return nil, fmt.Errorf("configured queue creation failed: %s", err)
+		}
+		// pull the properties from the parent that should be set on the child
+		sq.mergeProperties(parent.getProperties(), conf.Properties)
+	}
+	sq.UpdateSortType()
+
+	log.Logger().Debug("configured queue added to scheduler",
+		zap.String("queueName", sq.QueuePath))
+
+	return sq, nil
+}
+
+// Add a new queue to the system based on the placement rules
+// A dynamically added queue can never be the root queue so parent must be set
+// lock free as it cannot be referenced yet
+func NewDynamicQueue(name string, leaf bool, parent *Queue) (*Queue, error) {
+	// fail without a parent
+	if parent == nil {
+		return nil, fmt.Errorf("dynamic queue can not be added without parent: %s", name)
+	}
+	// name might not be checked do it here
+	if !configs.QueueNameRegExp.MatchString(name) {
+		return nil, fmt.Errorf("invalid queue name %s, a name must only have alphanumeric characters,"+
+			" - or _, and be no longer than 64 characters", name)
+	}
+	sq := newBlankQueue()
+	sq.Name = strings.ToLower(name)
```

Comment:
The regexp allows only lower case letters, we don't need this lower case conversion here.

Reply:
No it does not. The regexp should be:
```
var QueueNameRegExp = regexp.MustCompile(`^[a-zA-Z0-9_-]{1,64}$`)
```
The current regexp is not correct:
```
var QueueNameRegExp = regexp.MustCompile(`^[a-z-z0-9_-]{1,64}$`)
```
That was broken by YUNIKORN-335. It needs to be fixed, I'll fix the regexp and add some extra tests.

Reply:
I think it should not allow upper case letters, because since we are converting it internally to lower case ones, we already faced some issues that a queue with an upper case letter was deleted. See YUNIKORN-404

Reply:
Ignore my last comment. That was fixed on the partition side